File: sast-python2/case/completeness/single_app_tracing/class/subclass/subclass_002_F.py
# -*- coding: utf-8 -*-
# evaluation information start
# real case = false
# evaluation item = 完整度->单应用跟踪完整度->接口与类->子类对象
# scene introduction = 继承属性
# level = 2
# bind_url = completeness/single_app_tracing/class/subclass/subclass_002_F
# evaluation information end
import os


class BaseObject(object):
    def __init__(self, data):
        self.data = "safe_value"


class DerivedObject(BaseObject):
    pass


def subclass_002_F(taint_src):
    obj = DerivedObject(taint_src)
    taint_sink(obj.data)


def taint_sink(o):
    os.system(o)


if __name__ == u"__main__":
    taint_src = u"taint_src_value"
    subclass_002_F(taint_src)

File: sast-php/case/completeness/single_app_tracing/class/subclass_002_F.php
<?php
// evaluation information start
// real case = false
// evaluation item = 完整度->单应用追踪->类与对象
// scene introduction = 父类构造函数净化数据
// level = 2
// bind_url = completeness/single_app_tracing/class/subclass_002_F
// evaluation information end

class SafeBase {
    protected $data;

    public function __construct($val) {
        $this->data = "safe_init";
    }

    public function getData() {
        return $this->data;
    }
}

class SafeChild extends SafeBase {
    public function __construct($val) {
        parent::__construct($val);
    }
}

function subclass_002_F($__taint_src) {
    $obj = new SafeChild($__taint_src);
    __taint_sink($obj->getData());
}

function __taint_sink($o) {
    shell_exec($o);
}

$taint_src = "taint_src_value";
subclass_002_F($taint_src);
